S603 false positive on safe cmd

### Summary

I have cmd as list of constant strings, not example with `input()` as S603 doc suggest to verify, but it still flags as untrusted.

```python
import subprocess

git_diff = subprocess.run(["/usr/bin/git", "diff", "main", "--name-status"], check=True, capture_output=True)
```

```bash
$ ruff check --select S ~/ruff_bug.py
ruff_bug.py:3:12: S603 `subprocess` call: check for execution of untrusted input
```

### Version

0.11.2

--- a/crates/ruff_linter/resources/test/fixtures/flake8_bandit/S603.py
+++ b/crates/ruff_linter/resources/test/fixtures/flake8_bandit/S603.py
@@ -1,20 +1,34 @@
 from subprocess import Popen, call, check_call, check_output, run
 
 # Different Popen wrappers are checked.
-Popen("true", shell=False)
-call("true", shell=False)
-check_call("true", shell=False)
-check_output("true", shell=False)
-run("true", shell=False)
+a = input()
+Popen(a, shell=False)
+call(a, shell=False)
+check_call(a, shell=False)
+check_output(a, shell=False)
+run(a, shell=False)
 
 # Values that falsey values are treated as false.
-Popen("true", shell=0)
-Popen("true", shell=[])
-Popen("true", shell={})
-Popen("true", shell=None)
+Popen(a, shell=0)
+Popen(a, shell=[])
+Popen(a, shell={})
+Popen(a, shell=None)
 
 # Unknown values are treated as falsey.
-Popen("true", shell=True if True else False)
+Popen(a, shell=True if True else False)
 
 # No value is also caught.
-Popen("true")
+Popen(a)
+
+# Literals are fine, they're trusted.
+run("true")
+Popen(["true"])
+
+# Fine through assignments as well.
+run(c := "true")
+cmd = ["true"]
+run(cmd)
+
+# Imperfect literals cannot be trusted.
+cmd = ["echo", input()]
+run(cmd)

--- a/crates/ruff_linter/resources/test/fixtures/flake8_bandit/S603.py
+++ b/crates/ruff_linter/resources/test/fixtures/flake8_bandit/S603.py
@@ -8,7 +8,7 @@ check_call(a, shell=False)
 check_output(a, shell=False)
 run(a, shell=False)
 
-# Values that falsey values are treated as false.
+# Falsey values are treated as false.
 Popen(a, shell=0)
 Popen(a, shell=[])
 Popen(a, shell={})
@@ -23,6 +23,11 @@ Popen(a)
 # Literals are fine, they're trusted.
 run("true")
 Popen(["true"])
+Popen("true", shell=False)
+call("true", shell=False)
+check_call("true", shell=False)
+check_output("true", shell=False)
+run("true", shell=False)
 
 # Not through assignments though.
 cmd = ["true"]

--- a/crates/ruff_linter/resources/test/fixtures/flake8_bandit/S603.py
+++ b/crates/ruff_linter/resources/test/fixtures/flake8_bandit/S603.py
@@ -24,11 +24,13 @@ Popen(a)
 run("true")
 Popen(["true"])
 
-# Fine through assignments as well.
-run(c := "true")
+# Not through assignments though.
 cmd = ["true"]
 run(cmd)
 
-# Imperfect literals cannot be trusted.
-cmd = ["echo", input()]
-run(cmd)
+# Instant named expressions are fine.
+run(c := "true")
+
+# But non-instant are not.
+(e := "echo")
+run(e)

--- a/crates/ruff_linter/src/rules/flake8_bandit/mod.rs
+++ b/crates/ruff_linter/src/rules/flake8_bandit/mod.rs
@@ -104,6 +104,7 @@ mod tests {
     #[test_case(Rule::SuspiciousURLOpenUsage, Path::new("S310.py"))]
     #[test_case(Rule::SuspiciousNonCryptographicRandomUsage, Path::new("S311.py"))]
     #[test_case(Rule::SuspiciousTelnetUsage, Path::new("S312.py"))]
+    #[test_case(Rule::SubprocessWithoutShellEqualsTrue, Path::new("S603.py"))]
     fn preview_rules(rule_code: Rule, path: &Path) -> Result<()> {
         let snapshot = format!(
             "preview__{}_{}",

--- a/crates/ruff_linter/src/rules/flake8_bandit/rules/shell_injection.rs
+++ b/crates/ruff_linter/src/rules/flake8_bandit/rules/shell_injection.rs
@@ -310,19 +310,7 @@ pub(crate) fn shell_injection(checker: &Checker, call: &ast::ExprCall) {
                     }
                 }
                 // S603
-                Some(ShellKeyword {
-                    truthiness:
-                        Truthiness::False | Truthiness::Falsey | Truthiness::None | Truthiness::Unknown,
-                }) => {
-                    if checker.enabled(Rule::SubprocessWithoutShellEqualsTrue) {
-                        checker.report_diagnostic(Diagnostic::new(
-                            SubprocessWithoutShellEqualsTrue,
-                            call.func.range(),
-                        ));
-                    }
-                }
-                // S603
-                None => {
+                _ => {
                     if checker.enabled(Rule::SubprocessWithoutShellEqualsTrue) {
                         checker.report_diagnostic(Diagnostic::new(
                             SubprocessWithoutShellEqualsTrue,

--- a/crates/ruff_linter/src/rules/flake8_bandit/rules/shell_injection.rs
+++ b/crates/ruff_linter/src/rules/flake8_bandit/rules/shell_injection.rs
@@ -4,6 +4,7 @@ use ruff_diagnostics::{Diagnostic, Violation};
 use ruff_macros::{derive_message_formats, ViolationMetadata};
 use ruff_python_ast::helpers::Truthiness;
 use ruff_python_ast::{self as ast, Arguments, Expr};
+use ruff_python_semantic::analyze::typing::find_binding_value;
 use ruff_python_semantic::SemanticModel;
 use ruff_text_size::Ranged;
 
@@ -287,6 +288,30 @@ impl Violation for UnixCommandWildcardInjection {
     }
 }
 
+/// Check if an expression is a trusted input for subprocess.run.
+/// We assume that any str or list[str] literal
+/// (or variables directly assigned such literals) can be trusted.
+fn is_trusted_input(checker: &Checker, arg: &Expr) -> bool {
+    match arg {
+        Expr::StringLiteral(_) => true,
+        Expr::List(ruff_python_ast::ExprList { elts, .. }) => {
+            elts.iter().all(|elt| matches!(elt, Expr::StringLiteral(_)))
+        }
+        Expr::Named(ruff_python_ast::ExprNamed { value, .. }) => is_trusted_input(checker, value),
+        Expr::Name(name) => {
+            let semantic = checker.semantic();
+            let Some(binding) = semantic.only_binding(name).map(|id| semantic.binding(id)) else {
+                return false;
+            };
+            let Some(value) = find_binding_value(binding, semantic) else {
+                return false;
+            };
+            is_trusted_input(checker, value)
+        }
+        _ => false,
+    }
+}
+
 /// S602, S603, S604, S605, S606, S607, S609
 pub(crate) fn shell_injection(checker: &Checker, call: &ast::ExprCall) {
     let call_kind = get_call_kind(&call.func, checker.semantic());
@@ -311,11 +336,13 @@ pub(crate) fn shell_injection(checker: &Checker, call: &ast::ExprCall) {
                 }
                 // S603
                 _ => {
-                    if checker.enabled(Rule::SubprocessWithoutShellEqualsTrue) {
-                        checker.report_diagnostic(Diagnostic::new(
-                            SubprocessWithoutShellEqualsTrue,
-                            call.func.range(),
-                        ));
+                    if !is_trusted_input(checker, arg) {
+                        if checker.enabled(Rule::SubprocessWithoutShellEqualsTrue) {
+                            checker.report_diagnostic(Diagnostic::new(
+                                SubprocessWithoutShellEqualsTrue,
+                                call.func.range(),
+                            ));
+                        }
                     }
                 }
             }

--- a/crates/ruff_linter/src/rules/flake8_bandit/rules/shell_injection.rs
+++ b/crates/ruff_linter/src/rules/flake8_bandit/rules/shell_injection.rs
@@ -4,7 +4,6 @@ use ruff_diagnostics::{Diagnostic, Violation};
 use ruff_macros::{derive_message_formats, ViolationMetadata};
 use ruff_python_ast::helpers::Truthiness;
 use ruff_python_ast::{self as ast, Arguments, Expr};
-use ruff_python_semantic::analyze::typing::find_binding_value;
 use ruff_python_semantic::SemanticModel;
 use ruff_text_size::Ranged;
 
@@ -289,25 +288,14 @@ impl Violation for UnixCommandWildcardInjection {
 }
 
 /// Check if an expression is a trusted input for subprocess.run.
-/// We assume that any str or list[str] literal
-/// (or variables directly assigned such literals) can be trusted.
-fn is_trusted_input(checker: &Checker, arg: &Expr) -> bool {
+/// We assume that any str or list[str] literal can be trusted.
+fn is_trusted_input(arg: &Expr) -> bool {
     match arg {
         Expr::StringLiteral(_) => true,
         Expr::List(ruff_python_ast::ExprList { elts, .. }) => {
             elts.iter().all(|elt| matches!(elt, Expr::StringLiteral(_)))
         }
-        Expr::Named(ruff_python_ast::ExprNamed { value, .. }) => is_trusted_input(checker, value),
-        Expr::Name(name) => {
-            let semantic = checker.semantic();
-            let Some(binding) = semantic.only_binding(name).map(|id| semantic.binding(id)) else {
-                return false;
-            };
-            let Some(value) = find_binding_value(binding, semantic) else {
-                return false;
-            };
-            is_trusted_input(checker, value)
-        }
+        Expr::Named(named) => is_trusted_input(&named.value),
         _ => false,
     }
 }
@@ -336,7 +324,7 @@ pub(crate) fn shell_injection(checker: &Checker, call: &ast::ExprCall) {
                 }
                 // S603
                 _ => {
-                    if !is_trusted_input(checker, arg) {
+                    if !is_trusted_input(arg) {
                         if checker.enabled(Rule::SubprocessWithoutShellEqualsTrue) {
                             checker.report_diagnostic(Diagnostic::new(
                                 SubprocessWithoutShellEqualsTrue,

--- a/crates/ruff_linter/src/rules/flake8_bandit/rules/shell_injection.rs
+++ b/crates/ruff_linter/src/rules/flake8_bandit/rules/shell_injection.rs
@@ -292,7 +292,7 @@ impl Violation for UnixCommandWildcardInjection {
 fn is_trusted_input(arg: &Expr) -> bool {
     match arg {
         Expr::StringLiteral(_) => true,
-        Expr::List(ruff_python_ast::ExprList { elts, .. }) => {
+        Expr::List(ast::ExprList { elts, .. }) => {
             elts.iter().all(|elt| matches!(elt, Expr::StringLiteral(_)))
         }
         Expr::Named(named) => is_trusted_input(&named.value),
@@ -324,7 +324,7 @@ pub(crate) fn shell_injection(checker: &Checker, call: &ast::ExprCall) {
                 }
                 // S603
                 _ => {
-                    if !is_trusted_input(arg) {
+                    if !is_trusted_input(arg) || checker.settings.preview.is_disabled() {
                         if checker.enabled(Rule::SubprocessWithoutShellEqualsTrue) {
                             checker.report_diagnostic(Diagnostic::new(
                                 SubprocessWithoutShellEqualsTrue,

